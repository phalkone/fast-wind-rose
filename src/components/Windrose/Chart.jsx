/* eslint-disable react/jsx-fragments */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

/**
 * The chart that can be divided into a specified number of sectors
 */
function Chart (props) {
  return (
    <Fragment>
      {/* Sector lines */}
      {[...new Array(props.sectorCount).keys()].map(i => (
        <Fragment key={i}>
          <line
            x1={props.center}
            x2={props.center}
            y1={10}
            y2={props.center}
            stroke='grey'
            strokeWidth='1'
            transform={`rotate(${(0.5 + i) * (360 / props.sectorCount)},` +
                       ` ${props.center},${props.center})`}
          />
        </Fragment>
      ))}
      {/* Circles */}
      {[...new Array(4).keys()].map(i => (
        <Fragment key={i}>
          <circle
            cx={props.center}
            cy={props.center}
            r={(4 - i) * ((props.center - 10) / 4)}
            fill='none'
            strokeWidth='1'
            stroke='grey'
            opacity={i > 0 ? 0.5 : 1}
          />
        </Fragment>
      ))}
    </Fragment>
  )
}

Chart.propTypes = {
  /**
   * The number of sectors
   */
  sectorCount: PropTypes.number,
  /**
   * The center of the chart. x and y coordinates are the same.
   */
  center: PropTypes.number
}

export default Chart
