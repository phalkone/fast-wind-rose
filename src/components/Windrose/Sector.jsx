import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import colors from './Colors'

export default function Sector (props) {
  const speedCategories = []
  const center = 130
  const cos = Math.cos(Math.PI / 180 * (props.sectorSize / 2))
  const sin = Math.sin(Math.PI / 180 * (props.sectorSize / 2))
  let length = props.barLength

  for (let speed = 50; speed >= 0; speed -= 5) {
    const prev = speed === 50 ? 100 : speed + 5
    const count = props.speeds.filter((val) => val > speed && val <= prev).length
    if (count) {
      speedCategories.push([speed, length])
      length -= count * props.interval
    }
  }

  return (
    speedCategories.map((count) => (
      <Fragment key={count[0]}>
        <path
          d={`M ${center + (count[1] * sin)} ${center - (count[1] * cos)} A ${count[1]} ${count[1]}, 0, 0, 0, ${center - (count[1] * sin)} ${center - (count[1] * cos)} L 130 130 Z`}
          fill={colors[count[0]]}
          transform={`rotate(${props.sector * props.sectorSize}, 130, 130)`}
        />
      </Fragment>
    ))
  )
}

Sector.propTypes = {
  sector: PropTypes.number,
  sectorSize: PropTypes.number,
  barLength: PropTypes.number,
  speeds: PropTypes.array,
  interval: PropTypes.number
}
