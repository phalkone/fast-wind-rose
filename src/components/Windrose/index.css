button {
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}

button:hover {
  background-color: rgb(203, 203, 203);
}

select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

svg, select, button {
  font-family: Roboto, "Open Sans", sans-serif;
}
