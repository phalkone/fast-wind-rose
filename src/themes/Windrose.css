.windrose-button{
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}
 
.windrose-button:hover {
  background-color: rgb(203, 203, 203);
}

.windrose-select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

.windrose-container {
  font-family: Roboto, "Open Sans", sans-serif;
}