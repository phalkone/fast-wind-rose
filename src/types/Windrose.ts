export interface IWindrose {
  /**
   * Hide or display legend
   */
  legend: boolean,
  /**
   * Width/height of chart. Will be displayed in specified viewbox.
   */
  size: number,
  /**
   * Default number of sectors. Must be included in sectorArray.
   */
  sectorCount: number,
  /**
   * Fixed interval between data points specified in hours. For example if
   * data points are every 30 min, then value should be 0.5
   */
  interval: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string },
  /**
   * Array with directional data for the wind
   */
  dirData: object[],
  /**
  * Array with speed data for the wind
  */
  spdData: object[],
  /**
   * Key for speed data
   */
  spdKey: string,
  /**
   * Key for directional data
   */
  dirKey: string,
  /**
   * Scales the compilation scale against the viewbox. Choose a value lower
   * than 1 to scale down and larger than 1 to scale up.
   */
  enlarge: number,
  /**
   * Array that defines the number of sectors the user can choose from.
   */
  sectorArray: number[],
  /**
   * Common key between direction and speed data
   */
  commonKey: string
}

export interface ILegend {
  /**
   * Size of the chart. Width = Height.
   */
  size: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string }
}

export interface IShip {
  /**
   * The center of the ship path. x and y coordinates are assumed to be the same.
   */
  center: number
}

export interface IChart extends IShip {
  /**
   * The number of sectors
   */
  sectorSize: number
}

export interface IIntervalLabel extends IChart {
  /**
   * Defines which sector is being drawn. 0 being the first sector.
   */
  sector: number,
  /**
   * Interval between data points specified in hours
   */
  interval: number,
  /**
   * Array with the speeds that are part of this sector
   */
  speeds: number[]
}

export interface ISector extends IIntervalLabel, ILegend {
  /**
   * Total bar length of this sector. The sector with the largest interval will
   * have a bar length equal to the radius.
   */
  barLength: number,
  /**
   * Defines the bar length of 1 unit of interval.
   */
  unit: number,
}

export interface IToolTip {
  /**
   * x coordinate of top left corner of tooltip
   */
  x: number,
  /**
   * y coordinate of top left corner of tooltip
   */
  y: number,
  /**
   * text of tooltip
   */
  text: [string, number],
  /**
   * Scale factor due to viewBox of SVG vs actual size of SVG.
   */
  factor: number
}
