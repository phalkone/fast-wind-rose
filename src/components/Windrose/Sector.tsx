import React, { Fragment, useContext, useState } from 'react'
import { ToolTip } from './ToolTip'
import { IntervalLabel } from './IntervalLabel'
import { divideBySpeed } from '../../utils/Windrose.util'
import type { ISector } from '../../types/Windrose.types'
import { WindroseContext } from '.'

/**
 * Colored sectors as per the specified scale.
 */
export const Sector = (props: ISector) => {
  const context = useContext(WindroseContext)
  /* State used for displaying tooltip. Displayed when mouse enters sector. */
  const [entered, setEntered] = useState<[number, number, [string, number]]>([-1, 0, ['', 0]])

  /* Constants to avoid recalculation */
  const cos = Math.cos(Math.PI / 180 * (context.sectorSize / 2))
  const sin = Math.sin(Math.PI / 180 * (context.sectorSize / 2))

  /* Divide speeds as per specified scale in array */
  const speedCategories = divideBySpeed(props.speeds, context.scale, context.max,
    context.center, context.interval)

  return (
    <>
      {/* Show the sector divided as per speed categories */}
      {speedCategories.map((count) => (
        <Fragment key={count[0]}>
          <path
            d={`M ${(context.center + (count[1] * sin)).toFixed(2)} ` +
             `${(context.center - (count[1] * cos)).toFixed(2)} ` +
             `A ${count[1]} ${count[1]}, 0, 0, 0, ` +
             `${(context.center - (count[1] * sin)).toFixed(2)} ` +
             `${(context.center - (count[1] * cos)).toFixed(2)} ` +
             `L ${context.center} ${context.center} Z`}
            fill={context.scale[count[0]]}
            transform={`rotate(${props.sector * context.sectorSize}, ` +
                     `${context.center}, ${context.center})`}
            onMouseMove={(e) => setEntered([e.clientX, e.clientY, [count[2], count[3]]])}
            onMouseOut={() => setEntered([-1, 0, ['', 0]])}
          />
        </Fragment>
      ))}
      <IntervalLabel
          sector={props.sector}
          speeds={props.speeds}
      />
      {/* Display tooltip when mouse cursor entered the sector */}
      {entered[0] !== -1 &&
        <ToolTip
          x={entered[0]}
          y={entered[1]}
          text={entered[2]}
          factor={context.size / (context.center * 2)}
        />}
    </>
  )
}
