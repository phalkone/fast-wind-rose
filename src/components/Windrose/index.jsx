import React, { Fragment } from 'react'
import Legend from './Legend'
import Chart from './Chart'
import Ship from './Ship'
import Sector from './Sector'
import IntervalLabel from './IntervalLabel'
import PropTypes from 'prop-types'

export default class Windrose extends React.Component {
  constructor (props) {
    super(props)
    const sectors = [...new Array(360 / this.props.sectorSize).keys()].map((el) => [])
    for (let i = 0; i < this.props.data.length; i++) {
      let dir = this.props.data[i].direction + (this.props.sectorSize / 2)
      if (dir > 360) dir -= 360
      const cat = Math.floor(dir / this.props.sectorSize)
      sectors[cat].push(this.props.data[i].speed)
    }
    this.max = sectors.map((el) => el.length).sort((a, b) => b - a)[0]
    this.sectors = sectors
  }

  renderSectors () {
    return (
      this.sectors.map((speeds, i) => (
        <Fragment key={i}>
          {speeds.length &&
            <IntervalLabel
              sector={i}
              interval={speeds.length}
              sectorSize={this.props.sectorSize}
            />}
          {speeds.length &&
            <Sector
              sector={i}
              speeds={speeds}
              sectorSize={this.props.sectorSize}
              barLength={(speeds.length / this.max) * 120}
              interval={120 / this.max}
            />}
        </Fragment>
      ))
    )
  }

  render () {
    return (
      <svg
        version='1.1'
        viewBox='0 0 325 260'
        width={this.props.width}
        height={this.props.height}
      >
        {this.props.legend && <Legend />}
        <Chart sectorSize={this.props.sectorSize} />
        {this.renderSectors()}
        <Ship />
      </svg>
    )
  }
}

Windrose.propTypes = {
  height: PropTypes.number,
  width: PropTypes.number,
  sectorSize: PropTypes.number,
  legend: PropTypes.bool,
  data: PropTypes.array
}
