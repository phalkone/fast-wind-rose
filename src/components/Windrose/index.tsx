import React, { createContext, Fragment, useState } from 'react'
import { Legend } from './Legend'
import { Chart } from './Chart'
import { Ship } from './Ship'
import { Sector } from './Sector'
import { divideBySector } from '../../utils/Windrose.util'
import type { IWindrose, IWindroseContext } from '../../types/Windrose.types'
import '../../themes/Windrose.scss'

/**
 * Draws a windrose for provided relative wind data
 * @param {IWindrose} props Properties of the Windrose as descibed by IWindrose.
 */
const Windrose = (props: IWindrose) => {
  /* Use state to set the number of sectors and show/hide legend */
  const [sectorCount, setSectorCount] = useState<number>(props.sectorCount)
  const [legend, setLegend] = useState<boolean>(props.legend)

  /* Divide data points per sector */
  const sectors = divideBySector(sectorCount, props.dirData, props.dirKey,
    props.spdData, props.spdKey, props.commonKey)

  /* Set the viewBox size compared to the compilation size */
  const compilationSize = 260 / props.enlarge

  /* Add sectorCount if not part of sectorArray */
  if (!props.sectorArray.includes(sectorCount)) {
    props.sectorArray.push(sectorCount)
  }

  return (
    <div className='windrose-container'>
      {/* Selector for the number of sectors */}
      <select
        title='Number of sectors'
        value={sectorCount}
        className='windrose-select'
        onChange={e => { setSectorCount(Number(e.target.value)) }}
      >
        {props.sectorArray.map(s => (
          <Fragment key={s}>
            <option value={s}>{s}</option>
          </Fragment>
        ))}
      </select>
      <button
        title='Show/hide legend'
        className='windrose-button'
        onClick={e => { setLegend(!legend) }}
        style={{
          left: props.size - 30
        }}
      >{legend ? '<<' : '>>'}
      </button>
      {/* Windrose SVG with optional legend */}
      <svg
        version='1.1'
        viewBox={`0 0 ${compilationSize + (legend ? 65 : 0)} ${compilationSize}`}
        width={props.size + (legend ? 65 * (props.size / compilationSize) : 0)}
        height={props.size}
      >
        <WindroseContext.Provider value={{
          center: compilationSize / 2,
          scale: props.scale,
          sectorSize: 360 / sectorCount,
          interval: props.interval,
          size: props.size,
          max: sectors.map((el) => el.length).sort((a, b) => b - a)[0]
        }}>
          {/* Draw legend only if option is given as prop */}
          {legend && <Legend/>}
          <Chart/>
          {/* Draw each sector with interval label */}
          {sectors.map((speeds, i) => (
            speeds.length && <Fragment key={i}>
              <Sector
                sector={i}
                speeds={speeds}
              />
            </Fragment>
          ))}
          {/* Draw ship outline */}
          <Ship/>
        </WindroseContext.Provider>
        {/* Ensures that tooltips are always on top */}
        <use href='#tooltip' fill='black' />
        <use href='#tooltiptext' />
      </svg>
    </div>
  )
}

Windrose.defaultProps = {
  size: 260,
  spdKey: 'value',
  dirKey: 'value',
  commonKey: 'timestamp',
  sectorCount: 12,
  enlarge: 1,
  interval: 1,
  sectorArray: [4, 8, 12, 16, 24, 32, 36],
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  }
}

const defaultContext : IWindroseContext = {
  center: 130,
  scale: Windrose.defaultProps.scale,
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

export const WindroseContext = createContext(defaultContext)

export default Windrose
