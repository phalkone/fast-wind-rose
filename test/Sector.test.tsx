import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { Sector } from '../src/components/Windrose/Sector'
import { WindroseContext } from '../src/components/Windrose'
import type { IWindroseContext } from '../src/types/Windrose.types'
import renderer from 'react-test-renderer'

let container : HTMLElement
beforeEach(() => {
  container = document.createElement('svg')
  container.setAttribute('viewBox', '0 0 325 260')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
})

const context : IWindroseContext = {
  center: 130,
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  },
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

it('renders the correct number of speed categories', () => {
  act(() => {
    render(
      <Sector
        sector={0}
        speeds={[2, 2, 2, 7, 7, 7, 12, 12, 12, 17]}
      />
      , container)
  })
  expect(Array.from(container.children).filter((el : Element) => el.tagName === 'PATH').length).toEqual(5)
})

it('renders the correct rotation', () => {
  act(() => {
    render(
      <Sector
        sector={2}
        speeds={[2, 2, 2]}
      />
      , container)
  })
  const rotation = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'PATH')
    .map((el: Element) => el.getAttribute('transform'))
  expect(rotation).toEqual(['rotate(60, 130, 130)', 'rotate(60, 130, 130)'])
})

test('render a Sector', () => {
  const comp = renderer.create(
    <WindroseContext.Provider value={ context }>
      <Sector
        sector={0}
        speeds={[2, 2, 2, 7, 7, 7, 12, 12, 12]}
      />
    </WindroseContext.Provider>
  )
  const tree = comp.toJSON()
  expect(tree).toMatchSnapshot()
})
