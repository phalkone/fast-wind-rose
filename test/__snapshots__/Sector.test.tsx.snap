// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a Sector 1`] = `
Array [
  <path
    d="M 141.65 86.53 A 45 45, 0, 0, 0, 118.35 86.53 L 130 130 Z"
    fill="rgb(143,180,232)"
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    transform="rotate(0, 130, 130)"
  />,
  <path
    d="M 137.76 101.02 A 30 30, 0, 0, 0, 122.24 101.02 L 130 130 Z"
    fill="rgb(94,131,188)"
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    transform="rotate(0, 130, 130)"
  />,
  <path
    d="M 133.88 115.51 A 15 15, 0, 0, 0, 126.12 115.51 L 130 130 Z"
    fill="rgb(60,95,156)"
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    transform="rotate(0, 130, 130)"
  />,
  <path
    d="M 100.49 11.62 A 122 122, 0, 0, 1, 159.51 11.62"
    fill="none"
    id="curved0"
    transform="rotate(0, 130, 130)"
  />,
  <text
    fill="black"
    fontSize="11px"
    textAnchor="middle"
  >
    <textPath
      href="#curved0"
      startOffset="50%"
    >
      9.0h
    </textPath>
  </text>,
]
`;
